### PR TITLE
WIP: Start Open Source Documentation

### LINKED ISSUE #273: Community Versions Documentation and Implementation
I am aware there have been a number of Community version related issues eg #85 and #219 but I was wondering how you would feel about a PR adding some more documentation and maybe configuration to this stack.

As a user I would love to have it documented specifically that I can utilize community-only things by setting certain parameters. Also exclude certain packages from the install stack.
And I would love to be able to switch of the metrics-reporter from the inventory.

If you guys would be willing to get something like that merged I would draft a PR :)

### PR DESCRIPTION
**Pull request by Fobhep**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #273 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

**Test Configuration**:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Fobhep** [comment]
@JumaX @domenicbove  Linking you here for discussion.

So imho there are two things to be done to get a "none-commercial" version of the confluent stack.

1. Document open source usage and create an example inventory

2. Do not install any packages that are "commercial". I am a bit lost with all the packages confluent is offering, so maybe you can help me out here :)
If I am correct this is a list of all packages that are possibly installed within this repo.

- confluent-common :question: 
- confluent-rest-utils :question: 
- confluent-security :question:
- confluent-kafka :heavy_check_mark: 
- confluent-schema-registry  :heavy_check_mark: 
- confluent-kafka-connect-storage-common :heavy_check_mark: 
- confluent-kafka-connect-s3 :heavy_check_mark:  
- confluent-kafka-connect-elasticsearch :heavy_check_mark: 
- confluent-kafka-connect-jdbc :heavy_check_mark: 
- confluent-ksqldb :heavy_check_mark: 
- confluent-kafka-rest :heavy_check_mark: 
- confluent-server :x:
- confluent-control-center-fe :x:
- confluent-control-center :x:
- confluent-rebalancer :x: (assuming this is the auto data rebalancer)
- confluent-metadata-service :question: (not sure here - also #197 ) 
- confluent-kafka-connect-jms :x:
- confluent-kafka-connect-replicator :x:

Could you review this list and maybe also amend if there are any further steps I forgot about so far?

**Comment by domenicbove** [comment]
@Fobhep When I read your issue on this one, I thought you were asking us to install opensource kafka and configure it, not just install our free bits. I would need to ask some people internally. The big one I know of is confluent-kafka-2.12 vs confluent-server, the later being not free

**Comment by Fobhep** [comment]
From talking to various confluent people I always had the understanding that using the "open source and community (aka free) parts" was perfectly fine and also intended :) Maybe I am wrong there?

Also since those parts are partly documented https://docs.confluent.io/current/installation/available_packages.html

We can of course also change this PR completely and use not Confluent packages at all for an open source kafka installation - is this more what you excpected or are there any other possible scenarions you would see in the scope of this repository? :)

**Comment by domenicbove** [comment]
Hey! So I do not think we have plans to help with the install of opensource kafka, but I dont see a reason why helping install our community packages shouldnt be an option. 

I did try one thing which might answer your question on the packages. If those docs are correct, then running `yum install confluent-community-2.12` should get you the community packages and their dependencies. And heres an out put of that command:

```
Dependencies Resolved

=======================================================================================================================================
 Package                                                Arch                   Version                 Repository                 Size
=======================================================================================================================================
Installing:
 confluent-community-2.12                               noarch                 5.5.0-1                 Confluent                 3.6 k
Installing for dependencies:
 confluent-common                                       noarch                 5.5.0-1                 Confluent                 121 k
 confluent-kafka-2.12                                   noarch                 5.5.0-1                 Confluent                  57 M
 confluent-kafka-connect-elasticsearch                  noarch                 5.5.0-1                 Confluent                 2.2 M
 confluent-kafka-connect-jdbc                           noarch                 5.5.0-1                 Confluent                 8.1 M
 confluent-kafka-connect-s3                             noarch                 5.5.0-1                 Confluent                  22 M
 confluent-kafka-connect-storage-common                 noarch                 5.5.0-1                 Confluent                  40 M
 confluent-kafka-rest                                   noarch                 5.5.0-1                 Confluent                  47 M
 confluent-ksqldb                                       noarch                 5.5.0-1                 Confluent                  96 M
 confluent-rest-utils                                   noarch                 5.5.0-1                 Confluent                  19 M
 confluent-schema-registry                              noarch                 5.5.0-1                 Confluent                  83 M

Transaction Summary
=======================================================================================================================================
Install  1 Package (+10 Dependent packages)
```

Now, thinking about this PR. I think there will need to be more logic gates around certain properties/configs. Like i know c3 packages are included in some components so they can configure interceptors... ie if you dont install c3 packages then the interceptors code should be skipped.

**Comment by Fobhep** [comment]
hey @domenicbove Thanks for the feedback!
And a great catch - the community-package should be exactly what we need.

I see - thanks for the remark. I will try to dig into that and see what I can forge together :)

**Comment by Fobhep** [comment]
@domenicbove So I had a first idea how to deal with the package selection inside of `roles/confluent.variables_handlers/vars/main.yml`.

I am not sure how to best deal with a scenario where a user sets `confluent_community: true` but also `confluent_server_enabled: true`
That is why I introduced the "hard rest" in `roles/confluent.common/tasks/main.yml`.

File: non_commercial_hosts.yaml [hunk 1]
```diff
@@ -0,0 +1,365 @@
+---
```
Comment (domenicbove) [inline non_commercial_hosts.yaml:1]: Is this file necessary? how is it different from hosts_example?

File: roles/confluent.common/tasks/main.yml [hunk 2]
```diff
@@ -1,13 +1,12 @@
 ---
 # JVM Installation and Repo Setup
-- include_tasks: redhat.yml
-  when: ansible_os_family == "RedHat"
+- include_tasks: "{{ ansible_os_family }}.yml"
```
Comment (domenicbove) [inline roles/confluent.common/tasks/main.yml:3]: I like what you did here, but I think ansible_os_family resolves to RedHat and Debian not RedHat, Ubuntu, and Debian
Meaning the Ubuntu tasks will not be run

Im curious what ansible_distribution resolves to on a redhat/centos host, maybe that would work

File: roles/confluent.variables_handlers/vars/main.yml [hunk 3]
```diff
@@ -1,11 +1,13 @@
 ---
 zookeeper_service_name: confluent-zookeeper
 zookeeper_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka-2.12'}}"
-zookeeper_packages:
+zookeeper_community_packages:
```
Comment (domenicbove) [inline roles/confluent.variables_handlers/vars/main.yml:4]: One thing to make sure of here is these packages have a very specific order: dependencies first
Why this is important is: if you run `yum install confluent-server-5.4.0` it will install confluent-common-5.4.2 as a dependency...

My concern with this implementation is if you have confluent_community set to false then the union will put confluent-metadata service after zookeeper_main_package in the list.... I can do some testing. This might be a good time for the omit filter, I need to read more on that filter though

**Comment by domenicbove** [comment]
Can you make this PR into the master branch? we are moving away from adding features to already "released" branches. Master is going to become the 6.0 branch which will be cut later this summer

File: roles/confluent.variables_handlers/vars/main.yml (hunk 3 above)
Comment (domenicbove) [reply roles/confluent.variables_handlers/vars/main.yml:4]: update: omit filter did not work but this would:
```
zookeeper_packages:
  - confluent-common
  - confluent-rest-utils
  - "{{ '' if confluent_community|bool else 'confluent-metadata-service' }}"
  - "{{ zookeeper_main_package }}"
zookeeper_final_packages: "{{zookeeper_packages | difference(['']) }}"
```

File: non_commercial_hosts.yaml (hunk 1 above)
Comment (Fobhep) [reply non_commercial_hosts.yaml:1]: This is definitely not necessary - I will remove it

File: roles/confluent.common/tasks/main.yml (hunk 2 above)
Comment (Fobhep) [reply roles/confluent.common/tasks/main.yml:3]: I might have a solution - pushing it now

Comment (Fobhep) [reply roles/confluent.common/tasks/main.yml]: > Hello,
> 
> roles/confluent.common/tasks/main.yml
> 
> Dynamic inclusion of ansible_os_family.yaml tasks are not working and breaking the playbook. Tested it on Ubuntu and its not including the Ubuntu.yml

Thx - fair point. I implemented a different approach now

File: roles/confluent.variables_handlers/vars/main.yml (hunk 3 above)
Comment (Fobhep) [reply roles/confluent.variables_handlers/vars/main.yml:4]: Good catch! I will implement that

Comment (Fobhep) [reply roles/confluent.variables_handlers/vars/main.yml]: I used your approach but tried to keep the "API" stable, by not changing the final `zookeeper_package` variable

Comment (domenicbove) [reply roles/confluent.variables_handlers/vars/main.yml]: Looks good!

**Comment by domenicbove** [comment]
So another thing I noticed when trying this out, is the conditionals on certain properties need to be updating. For example the metrics reporter bit in the kafka_broker properties I don't think works when community is enabled.... I am not sure of the full list of properties (features) that should be turned off, but its all handled in variables_handlers/vars/main.yml

File: roles/confluent.common/tasks/CentOS.yml [hunk 4]
```diff
@@ -0,0 +1 @@
+RedHat.yml
```
Comment (domenicbove) [inline roles/confluent.common/tasks/CentOS.yml:1]: is this a symlink? unfortunately, some of our customers did some funky stuff with moving the source code around and it messed up symlinks when i added them before, so I think we are trying not to rely on them (bc we can't trust ppl's git skills)

To avoid symlinks, you should be able to make this a task file that just calls the red hat tasks file

**Comment by JumaX** [comment]
@domenicbove is correct regarding metrics reporter, this is proprietary to Confluent-Server and thus will not work with pure open source. It's used to supply metrics to Control Center.  This would apply to metrics configs across all components.

File: roles/confluent.common/tasks/CentOS.yml (hunk 4 above)
Comment (Fobhep) [reply roles/confluent.common/tasks/CentOS.yml:1]: ah too bad - but ok I will change that :)

**Comment by Fobhep** [comment]
> @domenicbove is correct regarding metrics reporter, this is proprietary to Confluent-Server and thus will not work with pure open source. It's used to supply metrics to Control Center. This would apply to metrics configs across all components.

ok thx - I will have a look at what I can find

**Comment by Fobhep** [comment]
@domenicbove @JumaX would the metrics reporter thing actually be "resolved" with #197 being merged and having a logic that triggers that feature when `confluent_community: true`

**Comment by Fobhep** [comment]
@domenicbove I finally had the chance to continue on this PR. I hope I caught all of the metrics reporters :)

File: roles/confluent.variables/vars/main.yml [hunk 5]
```diff
@@ -66,8 +66,8 @@ zookeeper_final_properties: "{{ zookeeper_combined_properties | combine(zookeepe
 
 
 #### Kafka Broker Variables ####
-kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka'}}"
-kafka_broker_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka-2.12'}}"
+kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool and not confluent_community|bool 'confluent-kafka'}}"
+kafka_broker_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool and not confluent_community|bool 'confluent-kafka-2.12'}}"
```
Comment (jrevillard) [inline roles/confluent.variables/vars/main.yml:70]:
```suggestion
kafka_broker_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool and not confluent_community|bool else 'confluent-kafka-2.12'}}"
```

File: roles/confluent.variables/vars/main.yml [hunk 6]
```diff
@@ -66,8 +66,8 @@ zookeeper_final_properties: "{{ zookeeper_combined_properties | combine(zookeepe
 
 
 #### Kafka Broker Variables ####
-kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka'}}"
-kafka_broker_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka-2.12'}}"
+kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool and not confluent_community|bool 'confluent-kafka'}}"
```
Comment (jrevillard) [inline roles/confluent.variables/vars/main.yml:69]:
```suggestion
kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool and not confluent_community|bool else 'confluent-kafka'}}"
```

**Comment by jrevillard** [comment]
I'm really interested in this PR is there any chance to get it reviewed and merged soon ?

Thanks for all your work.

Best,
Jérome

**Comment by Fobhep** [comment]
@jrevillard I will continue this hopefully later today or tomorrow by rebasing it on latest changes and putting in your fixes.
I will afterwards ping the repo maintainers to make sure we get some feedback and can continue from there on

**Comment by Fobhep** [comment]
@jrevillard  @JumaX @domenicbove  I updated this PR - please feel free to review.

Sadly I cannot check the failed Jenkins state - if somebody has an idea how I can fix that please let me know.
image

File: roles/confluent.common/tasks/main.yml [hunk 7]
```diff
@@ -16,14 +16,14 @@
...
-- include_tasks: redhat.yml
-  when: ansible_os_family == "RedHat"
 
-- include_tasks: ubuntu.yml
-  when: ansible_distribution == "Ubuntu"
+- include_tasks: "{{ ansible_distribution }}.yml"
 
-- include_tasks: debian.yml
-  when: ansible_distribution == "Debian"
+- name: Unset confluent features if community version
```
Comment (domenicbove) [inline roles/confluent.common/tasks/main.yml:22]: So this ensures that even when someone sets 
```
rbac_enabled: true
confluent_community: true
```
That rbac gets turned off. 

I am leaning toward putting this in the variables file and having an assertion, that complains thats not ok. WDYT?

Similar to the assertion in roles/confluent.control_center/tasks/main.yml

File: roles/confluent.variables/vars/main.yml [hunk 8]
```diff
@@ -65,15 +66,16 @@ zookeeper_final_properties: "{{ zookeeper_combined_properties | combine(zookeepe
 
 
 #### Kafka Broker Variables ####
-kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka'}}"
-kafka_broker_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka-2.12'}}"
-kafka_broker_packages:
+kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool and not confluent_community|bool else 'confluent-kafka'}}"
+kafka_broker_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool and not confluent_community|bool else 'confluent-kafka-2.12'}}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:70]:
```suggestion
kafka_broker_service_name: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka'}}"
kafka_broker_main_package: "{{ 'confluent-server' if confluent_server_enabled|bool else 'confluent-kafka-2.12'}}"
```

Since confluent_server_enabled should already be set correctly do you need the and condition?

File: roles/confluent.variables/defaults/main.yml [hunk 9]
```diff
@@ -39,6 +39,9 @@ open_file_limit: 500000
 ### Boolean to install commercially licensed confluent-server instead of community version: confluent-kafka
 confluent_server_enabled: true
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:40]:
```suggestion
### Boolean to install commercially licensed confluent-server instead of community version: confluent-kafka.
confluent_server_enabled: "{{ not confluent_community}}"
```
WDYT about setting these variables based off of the confluent_community variable? And then having assertions on them as well?

File: roles/confluent.common/tasks/main.yml [hunk 10]
```diff
@@ -16,14 +16,14 @@
     - ansible_distribution
 
 # JVM Installation and Repo Setup
-- include_tasks: redhat.yml
-  when: ansible_os_family == "RedHat"
 
-- include_tasks: ubuntu.yml
-  when: ansible_distribution == "Ubuntu"
+- include_tasks: "{{ ansible_distribution }}.yml"
```
Comment (domenicbove) [inline roles/confluent.common/tasks/main.yml:20]: I wonder if you could do:

```suggestion
- include_tasks: "{{ ansible_os_family if  ansible_os_family == 'RedHat' else ansible_distribution}}.yml"
```

And then we dont need the Centos.yml file

**Comment by nsharma-git** [comment]
@Fobhep Now that we have documentations, do you want to proceed with this PR? If yes, then can you please make it a regular PR and resolve the conflicts.

**Comment by nsharma-git** [comment]
IMO, this is taken care of. We can reopen if needed.

**Comment by nsharma-git** [comment]
IMO, this is taken care of. We can reopen if needed.
